auth/api/iam/s2s_vptoken.go

```diff
@@ -19,49 +19,157 @@
+const scopeJWTClaimKey = "scope"
+
+// scopeCredentialsClaimKey defines the JWT claim name for the Verifiable Credentials that were presented by the holder
+// when the token was created.
+const scopeCredentialsClaimKey = "vcs"
+
+// handleS2SAccessTokenRequest handles the /token request with vp_token bearer grant type, intended for service-to-service exchanges.
+// It performs cheap checks first (parameter presence and validity), then the more expensive ones (checking signatures and matching VCs to the presentation definition).
+func (s Wrapper) handleS2SAccessTokenRequest(issuer did.DID, params map[string]string) (HandleTokenRequestResponseObject, error) {
+	submissionEncoded := params["presentation_submission"]
+	scope := params[scopeParam]
+	assertionEncoded := params["assertion"]
+	if submissionEncoded == "" || scope == "" || assertionEncoded == "" {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "missing required parameters",
+		}
+	}
+
+	// Unmarshal VP, which can be in URL-encoded JSON or JWT format
+	if strings.HasPrefix(assertionEncoded, "ey") {
+		// (VP in JWT format)
+		return nil, errors.New("TODO: VPs in JWT format not yet supported")
+	}
+	// (VP as URL-encoded JSON)
+	assertionDecoded, err := url.QueryUnescape(assertionEncoded)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "assertion parameter is invalid: invalid query escaping",
+			InternalError: err,
+		}
+	}
+	var vp vc.VerifiablePresentation
+	err = json.Unmarshal([]byte(assertionDecoded), &vp)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "assertion parameter is invalid: invalid JSON",
+			InternalError: err,
+		}
+	}
+
+	// Unmarshal presentation submission
+	var submission pe.PresentationSubmission
+	submissionDecoded, err := url.QueryUnescape(submissionEncoded)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid query escaping",
+			InternalError: err,
+		}
+	}
+	if err = json.Unmarshal([]byte(submissionDecoded), &submission); err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid JSON",
+			InternalError: err,
+		}
+	}
 
-func (s serviceToService) validateVPToken(params map[string]string) (string, error) {
-	submission := params["presentation_submission"]
-	scope := params["scope"]
-	vp_token := params["vp_token"]
-	if submission == "" || scope == "" || vp_token == "" {
-		// TODO: right error response
-		return "", errors.New("missing required parameters")
-	}
-	// TODO: https://github.com/nuts-foundation/nuts-node/issues/2418
-	// TODO: verify parameters
-	return scope, nil
+	// Check signatures and VC issuer trust
+	issueTime := time.Now()
+	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
```

Comment:
`VerifyVP` does not check if the given credentials are owned by the client. RFC021 states:

| 7. The `verificationMethod` field of the proof MUST match the `credentialSubject.id` field from all the Verifiable Credentials that are used to request the access token.

This is not a general requirement for VP validation. We should hardcode it here. Also as discussed in front of the whiteboard, this check must be done before the discovery and cryptographic check to prevent DOS attacks.